LIN (#20)
hostmobility/documentation: docs/interfaces/mx4/lin.md

tags:
  - LIN
  - MX-4
  - T20
  - T30
  - T30 FR
  - V61
---

## Deactivate LIN
LIN is default on and you can deactivate it with sending a zero to
`LIN enabled` and/or `LIN2 enabled`. See [Digital IO](../digital_io.md) to use linux gpio.

## How to control LIN via serial interface

The LIN interface in the PIC is accessed via the serial console with baudrate 115200.

* T30 and T30 FR
    - LIN 1: /dev/ttyHS1
    - LIN 2: /dev/ttyHS3
* T20
    - LIN 1: /dev/ttyHS1
    - LIN 2: /dev/ttyHS3
* V61
    - LIN 1: /dev/ttyLP2
    - LIN 2: /dev/ttyLP1 (CRTSCTS must NOT be enabled or the communication will lock)

The LIN interface is controled via a set of predefined frames, mostly used to alter the LIN schedule table.

### All frames sent to co-cpu is on the format

```
start of transmission | message length | message type | data | checksum
```
### start of transmission (4 bytes)

This is a byte sequence used to indicate the start of a frame.
It consists of the 4 byte long sequence: ```0x7e 0x7e 0x7e 0xa7```

### message length (1 byte)

The length of message type (1 byte) + data (variable) + checksum (1 byte)
The byte for message length is not included in message length.

> 

Author: I agree most of it is from mx4-wiki. i will fix this.

### message type (1 byte)

Which type of operation that is requested.

> Values start at 1 and is sequentially increased.

Author: Yes, i am not sure how it works. The line is direct taken from mx4-wiki.

    data on the bus. LIN bus is default configured as master.
    When configured as listener all valid messages will be forwarded.
    Format:
    message start | length of message | message type 6 | data (master/listen 1
    byte) | checksum

    master - 1
    listen - 0
#### Listen message
    When the LIN bus is in listen mode messages will be sent to Linux user space when data are found on the bus.
    Format:
    message start | length of message | message type 7 | status| data | checksum

    status:
    Defined statuses:
```c
#define STATUS_OK 0
#define STATUS_RECV_ERROR 1
#define STATUS_RECV_NO_DATA 2
#define STATUS_MASTER_REQUEST 3
```
    STATUS_RECV_ERROR indicates that data have been read on the bus, but no valid LIN frame were found. Just all the data is returned for debuging purposes.
    STATUS_RECV_NO_DATA activity on the bus but no data found.
    STATUS_MASTER_REQUEST message with only one byte data found, this is a
    master request.

    data:
    The LIN frame.
    id | data | LIN checksum
    id: With parity omit bit 7 and 6 for id without parity
    LIN checksum: If it is a valid frame the checksum is here
#### Overflow message

>     When a overflow occurs at any of the busses handling LIN a report of which
>     bus and how many overflows that have occured is sent to the Linux system.

Reviewer: ```suggestion
    If an overflow occurs at any of the LIN buses, a report of the bus ID
    and the overflow count are sent to the Linux system.
```
